In [9]:
import numpy as np

labels = np.load('../data/train_labels.npy')
#data = np.load('../data/processed_train.npy')
data = np.load('../data/normalized.npy')
print(labels.shape, data.shape)

data = np.nan_to_num(data)

(19909,) (19909, 356)


In [10]:
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=19, random_state=0).fit(data)

unsuperlabels = kmeans.labels_
N = len(unsuperlabels)
unsuperlabels.shape


(19909,)

In [11]:
def clusters(label):
    res = []
    for i in range(19): 
        res.append([])

    for i in range(N):
        ind = label[i]
        res[ind].append(i)

    for i in range(19):
        res[i] = set(res[i])
        
    return res

def jaccard(s1,s2,p = False):
    inter = len(s1 & s2)
    union = len(s1 | s2)
    if p:
        print('inter: %d, union: %d'%(inter,union))
    return inter/union


In [12]:
origin = clusters(labels)
kmean = clusters(unsuperlabels)

jaccard_scores = []
for i in range(19):
    best = 0
    for j in range(19):
        jacc = jaccard(origin[i],kmean[j])
        if jacc > best:
            best = jacc
            print(i,'-',j, end = ' ')
    print()
    jaccard_scores.append(best)

jaccard_scores


0 - 0 0 - 2 0 - 18 
1 - 0 1 - 4 1 - 14 
2 - 0 2 - 3 2 - 4 2 - 14 
3 - 0 3 - 5 3 - 16 
4 - 0 4 - 1 4 - 2 
5 - 0 5 - 1 5 - 5 
6 - 0 6 - 2 6 - 18 
7 - 0 7 - 5 7 - 8 
8 - 0 8 - 1 
9 - 0 9 - 1 9 - 12 
10 - 0 10 - 2 10 - 16 
11 - 0 11 - 6 11 - 8 
12 - 0 12 - 1 12 - 6 12 - 8 12 - 12 
13 - 0 13 - 1 13 - 5 13 - 6 
14 - 0 14 - 1 14 - 5 
15 - 0 15 - 1 15 - 2 15 - 9 
16 - 0 16 - 1 16 - 9 
17 - 0 
18 - 0 18 - 1 18 - 2 18 - 3 18 - 9 


[0.1468189233278956,
 0.11982779239416408,
 0.14125376402956474,
 0.09800664451827243,
 0.19241672572643514,
 0.10827790096082779,
 0.06500351370344343,
 0.09940249864204237,
 0.05643232902917265,
 0.08532934131736528,
 0.057931034482758624,
 0.1434108527131783,
 0.29207920792079206,
 0.03027027027027027,
 0.016350210970464137,
 0.011255411255411256,
 0.016536118363794605,
 0.005904059040590406,
 0.003693444136657433]

In [5]:
def printclus(c):
    res = list(c)
    res.sort()
    print(res)
jaccard(origin[12],kmean[12],True)

inter: 236, union: 808


0.29207920792079206

In [7]:
data = np.load('../data/processed_train.npy')
data = np.nan_to_num(data)
kmeans = KMeans(n_clusters=19, random_state=0).fit(data)

unsuperlabels = kmeans.labels_
N = len(unsuperlabels)

origin = clusters(labels)
kmean = clusters(unsuperlabels)

jaccard_scores_unorm = []
for i in range(19):
    best = 0
    for j in range(19):
        jacc = jaccard(origin[i],kmean[j])
        if jacc > best:
            best = jacc
            print(i,'-',j, end = ' ')
    print()
    jaccard_scores_unorm.append(best)

jaccard_scores_unorm


0 - 0 
1 - 0 1 - 3 
2 - 0 2 - 3 2 - 15 
3 - 0 3 - 3 
4 - 0 4 - 12 4 - 14 
5 - 0 5 - 3 
6 - 0 6 - 1 
7 - 0 7 - 3 
8 - 0 
9 - 0 9 - 3 
10 - 0 
11 - 0 11 - 3 
12 - 0 12 - 3 
13 - 0 13 - 3 
14 - 0 14 - 3 
15 - 0 15 - 1 15 - 2 15 - 7 
16 - 3 
17 - 0 17 - 3 
18 - 0 18 - 5 18 - 12 18 - 14 


[0.10896164021164022,
 0.12025316455696203,
 0.12025654730090861,
 0.12877635296674636,
 0.09958333333333333,
 0.09520616828695944,
 0.057157569515962924,
 0.07572880598682007,
 0.04541305340027385,
 0.056054819141765896,
 0.040619863817797607,
 0.055923729798860594,
 0.04702595739727622,
 0.021596783457782882,
 0.009716599190283401,
 0.009433962264150943,
 0.0107113750145535,
 0.004181670344987803,
 0.0022396416573348264]

In [22]:
data = np.load('../data/norm_filtcorr.npy')
kmeans = KMeans(n_clusters=19, random_state=0).fit(data)

unsuperlabels = kmeans.labels_
N = len(unsuperlabels)

origin = clusters(labels)
kmean = clusters(unsuperlabels)

jaccard_scores_filt = []
for i in range(19):
    best = 0
    for j in range(19):
        jacc = jaccard(origin[i],kmean[j])
        if jacc > best:
            best = jacc
            print(i,'-',j, end = ' ')
    print()
    jaccard_scores_filt.append(best)

jaccard_scores_filt

0 - 0 0 - 1 
1 - 0 1 - 1 1 - 6 1 - 8 1 - 13 
2 - 1 2 - 6 2 - 13 
3 - 0 3 - 1 3 - 2 3 - 3 3 - 10 3 - 14 
4 - 1 4 - 2 4 - 5 
5 - 0 5 - 1 5 - 2 5 - 7 5 - 15 
6 - 0 6 - 1 6 - 5 6 - 9 
7 - 0 7 - 1 7 - 2 7 - 3 7 - 10 
8 - 0 8 - 1 8 - 2 8 - 7 
9 - 0 9 - 3 9 - 7 
10 - 0 10 - 1 
11 - 0 11 - 3 11 - 10 11 - 11 11 - 17 
12 - 0 
13 - 0 13 - 1 13 - 2 13 - 3 
14 - 2 14 - 7 14 - 8 
15 - 1 15 - 2 15 - 4 15 - 18 
16 - 0 16 - 7 
17 - 1 17 - 2 17 - 7 17 - 8 
18 - 1 18 - 2 18 - 3 18 - 9 


[0.137913741223671,
 0.1372960372960373,
 0.1860907362129856,
 0.0808931185944363,
 0.16193719258418462,
 0.15314401622718052,
 0.09848156182212581,
 0.10864197530864197,
 0.04963680387409201,
 0.08677966101694916,
 0.054520037278657965,
 0.1180327868852459,
 0.47326732673267324,
 0.0608,
 0.019933554817275746,
 0.01423785594639866,
 0.023174971031286212,
 0.006872852233676976,
 0.004418262150220913]

In [17]:
import pandas as pd
compare = [np.array(jaccard_scores),np.array(jaccard_scores_unorm)]
compare = np.array(compare).T
comp = pd.DataFrame(compare, columns = ['normalized', 'original'])

# Normalized v.s. Original 
|	|norm      |origin    |filter    |
|---|----------|----------|----------|
| 0	|`0.146819`|0.108962  |0.137913  |
| 1	|0.119828  |0.120253  |`0.137296`|
| 2	|0.141254  |0.120257  |`0.186090`|
| 3	|0.098007  |`0.128776`|0.080893  |
| 4	|`0.192417`|0.099583  |0.161937  |
| 5	|0.108278  |0.095206  |`0.153144`|
| 6	|0.065004  |0.057158  |`0.098481`|
| 7	|0.099402  |0.075729  |`0.108641`|
| 8	|`0.056432`|0.045413  |0.049636  |
| 9	|0.085329  |0.056055  |`0.086779`|
|10	|`0.057931`|0.040620  |0.054520  |
|11	|`0.143411`|0.055924  |0.118032  |
|12	|0.292079  |0.047026  |`0.473267`|
|13	|0.030270  |0.021597  |`0.0608`  |
|14	|0.016350  |0.009717  |`0.019933`|
|15	|0.011255  |0.009434  |`0.014237`|
|16	|0.016536  |0.010711  |`0.023174`|
|17	|0.005904  |0.004182  |`0.006872`|
|18	|0.003693  |0.002240  |`0.004418`|


Normalized features performed much better on unsupervised method (k-means)